File: homework/hw2/python/queens_conflict_minimize.py
from math import exp
from time import time

from problem.queens_constraint import QueensConstraintSatisfaction
from algorithm.conflicts_minimize import ConflictMinimize
from utils.selection import MaxSelection, FirstBetterSelection, RouletteSelection

if __name__ == '__main__':
    t0 = time()
    alpha = 10
    value_of = lambda value: exp(alpha * value)  # 以exp(10输入)作为估值 输入都是某行冲突数

    n = 1600
    q = QueensConstraintSatisfaction(n)
    cm = ConflictMinimize(q)

    fbs = FirstBetterSelection()
    rs = RouletteSelection()
    ms = MaxSelection()

    cm.search(10, n << 2, ms)

    print(f"Total time: {time() - t0}")

File: homework/hw2/c++/queens_conflict_minimize.cpp
#include <cmath>
#include <ctime>

#include "problem/queens_constraint.hpp"
#include "algorithm/conflicts_minimize.hpp"
#include "utils/selection.hpp"

// 在保证非负单调递增的前提下，估值函数仅对轮盘赌算法起作用
double alpha = 10;
double value_of(int value){
    return exp(alpha * value);
}

int main(){
    
    time_t t0 = time(nullptr);

    std::ios::sync_with_stdio(false);

    int n = 1000;
    
    QueensConstraintSatisfaction q(n);
    
    ConflictMinimize<QueensConstraintSatisfaction> cm(q);
    
    FirstBetterSelection fbs;
    RouletteSelection rs;
    MaxSelection ms;

    // 随机重启尝试10轮，每轮最多更改变元4n次
    // ms: 最大选择算法，优先选择冲突最多的变元更改，优先更改到冲突最小的值（移步algorithm/conflict_minimize.hpp阅读）
    // value_of: 因为使用最大选择算法，因此估值函数直接使用默认的指数函数即可
    cm.search(10, n << 2, ms);

    std::cout << "Total time: " << time(nullptr) - t0 << std::endl;
    return 0;
}